polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -1361,6 +1366,12 @@ impl<Config: config::Config> XcmExecutor<Config> {
 				result
 			},
 			PayFees { asset } => {
+				// If we've already paid for fees, do nothing.
+				if self.already_paid_fees {
```

Comment:
what about just filtering this by Barrier? Or do we want to be here like "best effort"?

Reply:
It could be with the barrier but this feels like something intrinsic to how the executor should work.
The barrier is meant for messages coming from other locations but this is something the executor wants to be able to always assume there's only one `PayFees` instruction executed